src/main/java/sprouts/impl/AbstractVariables.java

```diff
@@ -129,82 +128,110 @@ protected AbstractVariables( boolean isImmutable, Class<T> type, boolean allowsN
     @Override public final int size() { return _variables.size(); }
 
     /** {@inheritDoc} */
-    @Override public Vars<T> removeLast( int count )
-    {
-        if ( _isImmutable ) throw new UnsupportedOperationException( "This list is immutable." );
+    @Override
+    public Vars<T> removeLast( int count ) {
+        if ( _isImmutable )
+            throw new UnsupportedOperationException( "This list is immutable." );
+        if ( count < 0)
+            throw new IllegalArgumentException("Invalid count! Count must be non-negative.");
+
         count = Math.min( count, size() );
-        if ( count == 0 ) return this;
-        if ( count == 1 ) return removeLast();
-        for ( int i = 0; i < count; i++ ) _variables.remove( size() - 1 );
-        _triggerAction( Change.REMOVE, -1, null, null );
+
+        if ( count == 0 )
+            return this;
+        if ( count == 1 )
+            return removeLast();
+
+        Vars<T> vars = (Vars<T>) (_allowsNull ? Vars.ofNullable(_type) : Vars.of(_type));
+
+        List<Var<T>> subList = _variables.subList( size() - count, size() );
+        for ( Var<T> var : subList ) vars.add(var);
+        subList.clear();
+
+        _triggerAction( Change.REMOVE, size(), null, vars );
         return this;
     }
 
-    /**
-     *  Removes {@code count} number of properties from the end
-     *  of the list and returns them in a new list.
-     *  @param count The number of properties to remove.
-     *  @return A new list of properties.
-     */
-    @Override public Vars<T> popLast( int count )
-    {
-        if ( _isImmutable ) throw new UnsupportedOperationException( "This list is immutable." );
+    /** {@inheritDoc} */
+    @Override
+    public Vars<T> popLast( int count ) {
+        if ( _isImmutable )
+            throw new UnsupportedOperationException( "This list is immutable." );
+        if ( count < 0)
+            throw new IllegalArgumentException("Invalid count! Count must be non-negative.");
+
         count = Math.min( count, size() );
-        if ( count == 0 ) return Vars.of(type());
-        if ( count == 1 ) return Vars.of(popLast());
-        Vars<T> vars = Vars.of(type());
+
+        Vars<T> vars = (Vars<T>) (_allowsNull ? Vars.ofNullable(_type) : Vars.of(_type));
+
+        if ( count == 0 )
+            return vars;
+
         List<Var<T>> subList = _variables.subList( size() - count, size() );
         for ( Var<T> var : subList ) vars.add(var);
         subList.clear();
-        _triggerAction( Change.REMOVE, -1, null, null );
+        _triggerAction( Change.REMOVE, size(), null, vars );
```

Comment:
should it not be `_triggerAction( Change.REMOVE, size() - count, null, vars );`? 
Since that is the index at which the removal
of elements starts...

Reply:
In case of a clear, the index at which we start removing elements is 0.
So the range in which modification takes place is `0 - size()`.
The user can use the range `del.index() ... del.oldValues().size()` 
to update their data structure according to how the list changed.

If you use `index()` then it is the index at the end of and after the modification.
So the user has the reverse engineer the real index where the modification took place.

Reply:
Yea you are right. Wow, `subList.clear()` actually removes the items from
the original list the sub-list was taken from.
I did not expect that (always thought it was just a view...